### PR TITLE
Fix bug on multiplication overflow edge case

### PR DESCRIPTION
**Pull request by schouhy**
This PR fixes a bug on bound constraints in the multiplication of UnsignedInteger. It also adds a few tests for the edge cases.

### CODE REVIEW

**Review by ajgara — APPROVED** [review]

File: math/src/unsigned_integer/element.rs [hunk 1]
```diff
@@ -155,9 +155,13 @@ impl<const NUM_LIMBS: usize> Mul<&UnsignedInteger<NUM_LIMBS>> for &UnsignedInteg
...
             }
-            // 2.3
-            limbs[NUM_LIMBS - 1 - (i + n + 1)] = carry as u64;
+            if i + n + 1 < NUM_LIMBS {
+                // 2.3
+                limbs[NUM_LIMBS - 1 - (i + n + 1)] = carry as u64;
+                carry = 0;
+            }
         }
+        assert!(carry == 0, "UnsignedInteger multiplication overflow.");
```
Comment (Oppen) [inline math/src/unsigned_integer/element.rs:164]: Nit:
```suggestion
        assert_eq!(carry, 0, "UnsignedInteger multiplication overflow.");
```

**Review by Oppen — APPROVED** [review]

File: math/src/unsigned_integer/element.rs (hunk 1 above)
Comment (schouhy) [reply math/src/unsigned_integer/element.rs]: Thanks

**Review by jrchatruc — APPROVED** [review]
